### PR TITLE
fix transfer functions for multiple projects

### PR DESCRIPTION
**Pull request by gfr10598**
1. Remove misleading function names on right side of exports in enqueue.js
2. Parameterize project ID and destBucket in transfer.js, which handles early part of embargo.  This is intended to allow use in different projects.
3. Add project specializations of embargoOnFileNotification, for each of the three projects.

---
This change is [Reviewable](https://reviewable.io/reviews/m-lab/etl/212)

### CODE REVIEW

File: functions/embargo/transfer.js [hunk 1]
```diff
@@ -1,24 +1,48 @@
...
+ * appearing in scraper-mlab-oti:
+ *
+ * // Create the buckets
+   gsutil mb -p mlab-staging data-mlab-staging
+   gsutil mb -p mlab-staging functions-mlab-staging
+   // Deploy the functions.
+   gcloud beta functions deploy embargoOnFileNotificationStaging \
+     --stage-bucket=functions-mlab-staging \
+     --trigger-bucket=scraper-mlab-oti \
+     --project=mlab-staging
```
Comment (yachang) [inline functions/embargo/transfer.js]: No sandbox version?

**Comment by stephen-soltesz** [comment]
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[functions/embargo/transfer.js, line 25 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq9M_6XEm4qZ1TIG3Cq:-Kq9M_6XEm4qZ1TIG3Cr:b50bwbw) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L25)):*
> ```JavaScript
>  * To deploy this cloud function to mlab-oti (until we get autodeploy set up):
>  *
>  * // Create the buckets
> ```

These blocks could be simplified and clarified by using a variable, e.g.

```
GCLOUD_PROJECT=mlab-oti
gsutil mb -p $GCLOUD_PROJECT archive-$GCLOUD_PROJECT
...
```

Name translations e.g. mlab-oti to "Prod", make setup and configuration more brittle and ad-hoc and unlikely to be done consistently by different authors over time. This creates more opportunities for mistakes and wrong expectations.

I suggest deriving the function name suffix from GCLOUD_PROJECT also. I would not object to migrating everything to mlab-prod to avoid this continuing to come up as an issue.

```
SUFFIX=${GCLOUD_PROJECT##*-}
$ echo $SUFFIX 
oti
$ echo ${SUFFIX^}
Oti
# Or,
$ echo ${SUFFIX^^}
OTI
```

---

*[functions/embargo/transfer.js, line 44 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq9NBbcLzhqH-cZRI-p:-Kq9NBbcLzhqH-cZRI-q:b6b3kjh) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L44)):*
> ```JavaScript
>    gcloud beta functions deploy embargoOnFileNotificationStaging \
>      --stage-bucket=functions-mlab-staging \
>      --trigger-bucket=scraper-mlab-oti \
> ```

Is this the current staging configuration? Please add a note about what considerations need to be taken before enabling "DELETE". I imagine this making them more complex.

Is there a way to not require cross-project configuration?

---

*[functions/embargo/transfer.js, line 93 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq9Oiidtv3vYIgmiTpx:-Kq9Oiidtv3vYIgmiTpy:b-vcu4sl) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L93)):*
> ```JavaScript
>  * Create a function to copy and delete a single file.
>  * Ideally, the destination should be determined based on the project ID,
>  * but the project ID does not seem to be reliably available.
> ```

If we follow the convention that buckets should be suffixed with the project name, can we use this to extract the project name?

---

*[functions/embargo/transfer.js, line 176 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq9SimHcEbp1Z7mnel2:-Kq9SimHcEbp1Z7mnel3:b-pk9vdt) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L176)):*
> ```JavaScript
>  * @param {function} done The callback function called when this function completes.
>  */
> exports.embargoOnFileNotification = function (event, project, destBucket, done) {
> ```

I don't understand javascript -- it looks like project is unused in this function. What am I missing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/212)*

**Comment by gfr10598** [comment]
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[functions/embargo/transfer.js, line 25 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq9M_6XEm4qZ1TIG3Cq:-KqO6TQ1eTTouzujJm42:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L25)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

These blocks could be simplified and clarified by using a variable, e.g.

```
GCLOUD_PROJECT=mlab-oti
gsutil mb -p $GCLOUD_PROJECT archive-$GCLOUD_PROJECT
...
```

Name translations e.g. mlab-oti to "Prod", make setup and configuration more brittle and ad-hoc and unlikely to be done consistently by different authors over time. This creates more opportunities for mistakes and wrong expectations.

I suggest deriving the function name suffix from GCLOUD_PROJECT also. I would not object to migrating everything to mlab-prod to avoid this continuing to come up as an issue.

```
SUFFIX=${GCLOUD_PROJECT##*-}
$ echo $SUFFIX 
oti
$ echo ${SUFFIX^}
Oti
# Or,
$ echo ${SUFFIX^^}
OTI
```

</blockquote></details>

Done.

---

*[functions/embargo/transfer.js, line 44 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq9NBbcLzhqH-cZRI-p:-KqO6qNG-tERZtTSYfIP:b54a9xa) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L44)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Is this the current staging configuration? Please add a note about what considerations need to be taken before enabling "DELETE". I imagine this making them more complex.

Is there a way to not require cross-project configuration?
</blockquote></details>

There is not currently a staging or sandbox config.  This change is intended to enable that, so that Ya can proceed with testing for embargo.

Do you mean multi-project config?  This is the best I've come up with?
If you mean configs that touch buckets in multiple projects, I don't see a good way around that either.

---

*[functions/embargo/transfer.js, line 45 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq8t-FuwSdhfswc7mCA-r1-45:-Kq9Z55-nidMqtJ3a5R5:bve5ado) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L45)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

No sandbox version?
</blockquote></details>

Added sandbox.

---

*[functions/embargo/transfer.js, line 93 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq9Oiidtv3vYIgmiTpx:-Kq9YsCagt8xeoh50wak:b-328hpi) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L93)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

If we follow the convention that buckets should be suffixed with the project name, can we use this to extract the project name?
</blockquote></details>

I think that would be good to do, but reluctant to do it now.  I like the redundancy so that we are less likely to accidentally configure something running mlab-oti.

---

*[functions/embargo/transfer.js, line 176 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq9SimHcEbp1Z7mnel2:-Kq9Ygaf4AoLEJMFK-KC:b-sz0ab) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L176)):*
> name
It is now unused.  I included it because Ya is about to use it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/212)*

**Comment by stephen-soltesz** [comment]
:lgtm:

A few typos and a question about whether we can normalize the mlab-oti and mlab-staging configs without adding special-case dependencies on scraper-mlab-oti.

---

Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[functions/embargo/transfer.js, line 44 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq9NBbcLzhqH-cZRI-p:-KqOohTuHM58d9NvpaJJ:b8as5x3) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L44)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

There is not currently a staging or sandbox config.  This change is intended to enable that, so that Ya can proceed with testing for embargo.

Do you mean multi-project config?  This is the best I've come up with?
If you mean configs that touch buckets in multiple projects, I don't see a good way around that either.



</blockquote></details>

In this case, my understanding is that we're trying to funnel more data through the mlab-staging deployment to stress it. Is that correct?

I was trying to ask, is there a way to setup and test mlab-staging without depending on mlab-oti? Cross project dependencies make things more complex. Depending on "more production" things is preferable to depending on "less production" things (e.g. mlab-oti depending on something running in mlab-staging), but not adding cross project dependencies would be preferable to both, imo.

For example, could we schedule cloud storage transfers between the archive-mlab-oti and scraper-mlab-staging (iiuc) so that the full staging config (including delete) could operate as-if scraper were adding those archives to scraper-mlab-staging?

Do I have the full picture?

---

*[functions/embargo/transfer.js, line 32 at r2](https://reviewable.io:443/reviews/m-lab/etl/212#-KqOneaEv66B4IO0BquS:-KqOneaF1DghxYjtz8b5:bbsd0o6) ([raw file](https://github.com/m-lab/etl/blob/c60fd82d0b9c3ca70dcbef79aaf086ace60623dd/functions/embargo/transfer.js#L32)):*
> ```JavaScript
>    // Deploy the functions.
>    export FN_SUFFIX=${GCLOUD_PROJECT##*-}
>    gcloud beta functions deploy embargoOnFileNotification${SUFFIX^} \
> ```

mismatch between FN_SUFFIX & SUFFIX.

---

*[functions/embargo/transfer.js, line 55 at r2](https://reviewable.io:443/reviews/m-lab/etl/212#-KqOnl0Gn09V8PQuBXKx:-KqOnl0Gn09V8PQuBXKy:b84wf2b) ([raw file](https://github.com/m-lab/etl/blob/c60fd82d0b9c3ca70dcbef79aaf086ace60623dd/functions/embargo/transfer.js#L55)):*
> ```JavaScript
>    gsutil mb -p $GCLOUD_PROJECT functions-$GCLOUD_PROJECT
>    // Deploy the functions.
>    gcloud beta functions deploy embargoOnFileNotification${SUFFIX^} \
> ```

Does SUFFIX want to be reset here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/212#-:-KqOqi6Fkk3gzMHfZbcP:bvaumwt)*

**Comment by gfr10598** [comment]
Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[functions/embargo/transfer.js, line 44 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq9NBbcLzhqH-cZRI-p:-KqP-VkBQt5d8XmkikXB:b-5bbs8s) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L44)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

In this case, my understanding is that we're trying to funnel more data through the mlab-staging deployment to stress it. Is that correct?

I was trying to ask, is there a way to setup and test mlab-staging without depending on mlab-oti? Cross project dependencies make things more complex. Depending on "more production" things is preferable to depending on "less production" things (e.g. mlab-oti depending on something running in mlab-staging), but not adding cross project dependencies would be preferable to both, imo.

For example, could we schedule cloud storage transfers between the archive-mlab-oti and scraper-mlab-staging (iiuc) so that the full staging config (including delete) could operate as-if scraper were adding those archives to scraper-mlab-staging?

Do I have the full picture?
</blockquote></details>

Yes that is absolutely possible.  I'm adding another comment about that.

---

*[functions/embargo/transfer.js, line 32 at r2](https://reviewable.io:443/reviews/m-lab/etl/212#-KqOneaEv66B4IO0BquS:-KqP-IK_ji00dZwFmhm8:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/c60fd82d0b9c3ca70dcbef79aaf086ace60623dd/functions/embargo/transfer.js#L32)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

mismatch between FN_SUFFIX & SUFFIX.
</blockquote></details>

Done.

---

*[functions/embargo/transfer.js, line 55 at r2](https://reviewable.io:443/reviews/m-lab/etl/212#-KqOnl0Gn09V8PQuBXKx:-KqP-I-z1nsN5KtSSQds:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/c60fd82d0b9c3ca70dcbef79aaf086ace60623dd/functions/embargo/transfer.js#L55)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Does SUFFIX want to be reset here?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/212)*

**Comment by gfr10598** [comment]
Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[functions/embargo/transfer.js, line 44 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq9NBbcLzhqH-cZRI-p:-KqP0hZgeELrEKq6FBng:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L44)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Yes that is absolutely possible.  I'm adding another comment about that.
</blockquote></details>

Done.

---

*[functions/embargo/transfer.js, line 45 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq8t-FuwSdhfswc7mCA-r1-45:-KqP0hwnAGy_G707yKoC:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L45)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Added sandbox.
</blockquote></details>

Done.

---

*[functions/embargo/transfer.js, line 32 at r2](https://reviewable.io:443/reviews/m-lab/etl/212#-KqOneaEv66B4IO0BquS:-KqP0dJuJwhlPUb89H5D:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/c60fd82d0b9c3ca70dcbef79aaf086ace60623dd/functions/embargo/transfer.js#L32)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---

*[functions/embargo/transfer.js, line 55 at r2](https://reviewable.io:443/reviews/m-lab/etl/212#-KqOnl0Gn09V8PQuBXKx:-KqP0egOkPrp_fbbklS4:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/c60fd82d0b9c3ca70dcbef79aaf086ace60623dd/functions/embargo/transfer.js#L55)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/212)*

**Comment by gfr10598** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[functions/embargo/transfer.js, line 45 at r1](https://reviewable.io:443/reviews/m-lab/etl/212#-Kq8t-FuwSdhfswc7mCA-r1-45:-KqPNHaIlxOECbCR8w9H:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/59c93eba156590269f2d575eb8c15cc7a0c9bf00/functions/embargo/transfer.js#L45)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/212)*

**Pull request merged**